### PR TITLE
feat!: Add `AlterField` Operation to Migration Generator

### LINKED ISSUE #205: Support "alter field" in automatic migration generator
We should support changing metadata of existing fields in the automatic migration generator. In other words, if someone changes some properties about a database field without changing its name, we should be able to migrate this automatically.

The entry point for this is https://github.com/cot-rs/cot/blob/2bbb09699bf7b74c41bb323338a02c528de1acff/cot-cli/src/migration_generator.rs#L391. This will also require adding a new `OperationInner` variant: https://github.com/cot-rs/cot/blob/2bbb09699bf7b74c41bb323338a02c528de1acff/cot/src/db/migrations.rs#L460

Note that the "alter field" operation will need the old version of the fields modified, too, so that an implementation of [`Operation::backwards`](https://github.com/cot-rs/cot/blob/2bbb09699bf7b74c41bb323338a02c528de1acff/cot/src/db/migrations.rs#L438) can be created.

### PR DESCRIPTION
**Pull request by kumarUjjawal**
## Overview

This PR introduces support for the `AlterField` operation in the migration generator. 

## Related Issues

#205 

## Reviewer Notes

- I don't have full confidence in my implementation if you think this is not what was the intention of the #205 feel free to close this for now.

### CODE REVIEW

**Comment by kumarUjjawal** [comment]
@seqre any feedback for me?

**Review by seqre — CHANGES_REQUESTED** [review]
This operation has many edge cases, and it should be thoroughly tested. Can you please add more tests first? 

Think about possible cases, such as: name change (this one can be skipped for now, it's tricky), type change, attributes change, etc. All those should be either handled or visibly err when a given case is unhandled or ambiguous. 

It seems that `cot-cli/src/migration_generator.rs` does not have a test for alter model, would you be so kind and add that too?

**Comment by kumarUjjawal** [comment]
Sure, Thank you!

**Comment by kumarUjjawal** [comment]
@m4tx the error doesn't seem to come from my PR, can you take a look?

File: cot-cli/src/migration_generator.rs
Comment (seqre) [inline cot-cli/src/migration_generator.rs:1414]: This enum should have been marked as `#[non_exhaustive]`, can you do this please?

File: cot-cli/src/migration_generator.rs [hunk 1]
```diff
@@ -1130,24 +1125,22 @@ impl GeneratedMigration {
...
+                    because it doesn't create a new model"
+                    ),
+                    DynOperation::RemoveModel { .. } => unreachable!(
+                        "RemoveModel operation shouldn't be a dependency of CreateModel \
+                    because it doesn't create a new model"
+                    ),
+                    DynOperation::AlterField { .. } => unreachable!(
+                        "AlterField operation shouldn't be a dependency of CreateModel \
+                    because it doesn't create a new model"
+                    ),
```
Comment (seqre) [inline cot-cli/src/migration_generator.rs]: I think we should just make a blanket default response, DynOperation will grow over time and it'd be annoying to add more entries to such match statements. The same would apply for lines 1174-1184. What do you think @m4tx?

Or should we just redo `DynOperation` so that all of those actions are just trait methods that we just call here? lol

**Review by seqre — COMMENTED** [review]
Thank you for your patience. I fixed the PR so it compiles and the tests pass! I left two comments and I'll try to test it myself tonight.

File: cot-cli/src/migration_generator.rs
Comment (kumarUjjawal) [reply cot-cli/src/migration_generator.rs:1414]: I should be able to make the changes on weekends. Thanks for the fix.

File: cot-cli/src/migration_generator.rs [hunk 2]
```diff
@@ -828,20 +828,15 @@ impl MigrationOperationGenerator {
...
-
-        #[expect(unreachable_code)]
-        print_status_msg(
-            StatusType::Modified,
-            &format!(
-                "Field '{}' from Model '{}'",
-                &migration_field.name, migration_model.model.name
+                &migration_field.name, app_model.model.name
             ),
         );
```
Comment (seqre) [inline cot-cli/src/migration_generator.rs:838]: Please bring this back

File: cot/src/db/migrations.rs [hunk 3]
```diff
@@ -521,6 +526,17 @@ impl Operation {
...
+                table_name,
+                old_field: _,
+                new_field,
+            } => {
+                let query = sea_query::Table::alter()
+                    .table(*table_name)
+                    .modify_column(new_field.as_column_def(database))
+                    .to_owned();
+                database.execute_schema(query).await?;
+            }
```
Comment (seqre) [inline cot/src/db/migrations.rs:592]: Could you please write some tests that cover this part?

File: cot/src/db/migrations.rs [hunk 4]
```diff
@@ -594,6 +610,18 @@ impl Operation {
...
+                old_field,
+                new_field: _,
+            } => {
+                // To reverse an alteration, set the column back to the old definition
+                let query = sea_query::Table::alter()
+                    .table(*table_name)
+                    .modify_column(old_field.as_column_def(database))
+                    .to_owned();
+                database.execute_schema(query).await?;
+            }
```
Comment (seqre) [inline cot/src/db/migrations.rs:683]: Could you please write some tests that cover this part?

File: cot-cli/src/migration_generator.rs [hunk 5]
```diff
@@ -1282,6 +1254,18 @@ impl GeneratedMigration {
...
+                    model_ty,
+                    ..
+                } => {
+                    let mut ops = vec![(i, model_ty.clone())];
+                    // Only depend on the new foreign key, not the old one
+                    if let Some(to_type) = foreign_key_for_field(new_field) {
+                        ops.push((i, to_type));
+                    }
+                    ops
+                }
```
Comment (seqre) [inline cot-cli/src/migration_generator.rs:1273]: Could you please write some tests that cover this part?

This one interests me, from what I've read this would only apply when changing constraint on the foreign key, but that's a 3-operation situation (remove constraint, modify, add constraint), which I'm not sure would be supported in our current setup.

File: cot/src/db/migrations.rs [hunk 6]
```diff
@@ -430,6 +430,11 @@ impl Operation {
         RemoveModelBuilder::new()
     }
 
+    // TODO: docs
```
Comment (seqre) [inline cot/src/db/migrations.rs]: Could you please write the docs in similar fashion as other items above?

File: cot/src/db/migrations.rs [hunk 7]
```diff
@@ -1667,6 +1700,51 @@ impl CustomBuilder {
     }
 }
 
+// TODO: docs
```
Comment (seqre) [inline cot/src/db/migrations.rs]: Could you please write the docs in similar fashion as other items above?

**Review by seqre — CHANGES_REQUESTED** [review]
Hey @kumarUjjawal, it seems you force-pushed your old changes which broke tests and removed the custom operations added recently to `master`. I've fixed those for you, please remember to `git pull` before doing any changes!

Regarding the PR, we're getting closer with every iteration! This feature is really important, but it's slightly complicated, that's why I'm pushing for it to be well tested. I appreciate your efforts!

**Comment by seqre** [comment]
Hey @kumarUjjawal, how's the work going on this PR? Do you need any help from us?

**Comment by kumarUjjawal** [comment]
Hi @seqre Sorry for the delay, I got busy with another project, doesn't seem like I can give time for the next few weeks Can you pick this up from here?

**Comment by m4tx** [comment]
> Hi @seqre Sorry for the delay, I got busy with another project, doesn't seem like I can give time for the next few weeks Can you pick this up from here?

Hey @kumarUjjawal, sure, that's alright! I've tried to pick it up and found out that the implementation of this for SQLite will require some more complex work, as this database engine doesn't support the `ALTER FIELD` syntax. Other frameworks typically emulate this by a "create-copy-drop-rename" table dance, but it's not trivial to do this with Cot currently as it would require providing some more context to the migration engine.

I think we'll take it from here and will be able to finish this; thanks for laying out the groundworks for this!
